### PR TITLE
Question selection features

### PR DESCRIPTION
**Pull request by asif-khan17**
**Fixed Issues**

#21172, #21173

**Files Changed**
TestController.cs - GetTestCategoryQuestions() , AddTestQuestion() , GetTestDetailsById().
TestRepository.cs: AddTestQuestionsAsync() method added, to add questions to test
GetAllTestCategoryQuestionsByIdAsync() method added to get all questions from
particular category by its Id
GetTestDetailsByIdAsync() added to get details of a particular test by its Id.
test-question.component.ts : getAllQuestion(), getTestDetails(), addTestQuestions().
test.routing.ts,question.model.ts, category.model.ts,test.service.ts. test-question.html
Files Added: TestAC,  CategoryAC, Test-Question-Selection(Migration),
test-category.model, test-questions.model, test-category.model,

**Checks**

- [X] Naming Conventions
- [X] Server side code comments (proper English, grammar and no spelling mistakes)
- [X] Client side code comments (proper English, grammar and no spelling mistakes)
- [X] Unused variables, methods, blank spaces and name spaces. 
- [X] Optimal Code and code formatting
- [X] Commit History is proper, no merge is done. 
- [X] Commit/pull request messages should be proper to justify your feature
- [X] All test cases are executed provided by tester.

**Known Issues**

**Comments**

NA

---
This change is [Reviewable](https://reviewable.io/reviews/promact/trappist/291)

### CODE REVIEW

File: Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs [hunk 1]
```diff
@@ -171,5 +191,153 @@ private DomainModel.Models.TestConduct.TestAttendees TestAttendee()
...
+                question.IsSelect = true;
+                questionAC.Question = question;
+                questionListAC.Add(questionAC);
+            }
+     
+            await  _testRepository.AddTestQuestionsAsync(questionListAC, test.Id);
+            var questionACList = await _testRepository.GetAllTestCategoryQuestionsByIdAsync(test.Id,category1.Id);
+            Assert.Equal(1,questionACList.Count);
+            Assert.Equal(2, _trappistDbContext.TestQuestion.Count());
+            Assert.True(questionACList[0].Question.IsSelect == true);
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

File: Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs [hunk 2]
```diff
@@ -171,5 +191,153 @@ private DomainModel.Models.TestConduct.TestAttendees TestAttendee()
...
+            return category;
+        }
+        /// <summary>
+        ///Test case for  getting test details containing categories only
+        /// </summary>
+        /// <returns></returns>
+        [Fact]
+        public async Task GetTestDetails()
+        {
+            TestAC testAC = new TestAC();
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs]: MAJOR Remove this useless assignment to local variable 'testAC'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1854)

File: Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestAC.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Promact.Trappist.DomainModel.ApplicationClasses.Test
+{
+    public class TestAC
+    {
+        public int Id { get; set; }
+       
+        public string TestName { get; set; }
+       
+        public List<CategoryAC> CategoryACList { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestAC.cs]: MINOR Rename property 'CategoryACList' to match camel case naming rules, consider using 'CategoryAcList'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 4]
```diff
@@ -101,5 +116,77 @@ public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util)
...
+            return questionListAC;
+        }
+
+        public async Task<string> AddTestQuestionsAsync(List<QuestionAC> QuestionsToAddTest,int testId)
+        {
+            List<TestQuestion> testQuestionList = new List<TestQuestion>();
+            foreach (var questionToAdd in QuestionsToAddTest)
+            {
+                var question = Mapper.Map<QuestionDetailAC, Question>(questionToAdd.Question);
+                var TestCategory = _dbContext.TestCategory.FirstOrDefault(x=>x.CategoryId==question.CategoryID);
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: MAJOR Remove this unused 'TestCategory' local variable. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1481)

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 5]
```diff
@@ -15,10 +22,18 @@ public class TestsRepository : ITestsRepository
     {
         private readonly TrappistDbContext _dbContext;
         private readonly IGlobalUtil _util;
-        public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util)
+        private readonly IStringConstants _stringConstants;
+        private QuestionAC questionAc;
+        private TestAC testACObject;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: MAJOR Remove the 'testACObject' field and declare it as a local variable in the relevant methods. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1450)

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 6]
```diff
@@ -15,10 +22,18 @@ public class TestsRepository : ITestsRepository
...
+        private TestQuestion testQuestionObj;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: MAJOR Remove the 'testQuestionObj' field and declare it as a local variable in the relevant methods. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1450)

**Comment by chintans** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs [hunk 7]
```diff
@@ -162,6 +165,28 @@ public TestsController(ITestsRepository testRepository, IStringConstants stringC
             await _testRepository.DeleteTestAsync(id);
             return NoContent();
         }
+
+        [HttpGet("questions/{testid}/{categoryid}")]
+        public async Task<IActionResult> GetTestCategoryQuestions([FromRoute] int  testid,[FromRoute] int categoryid)
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs]: Naming conventions are not proper

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 8]
```diff
@@ -101,5 +111,78 @@ public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util)
...
+            });
+            return questionListAc;
+        }
+
+        public async Task<string> AddTestQuestionsAsync(List<QuestionAC> questionsToAddTest, int testId)
+        {
+            bool isDeleted = false;
+            TestQuestion testQuestionObj = new TestQuestion();
+            List<TestQuestion> testQuestionList = new List<TestQuestion>();
+            foreach (var questionToAdd in questionsToAddTest)
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: MAJOR Remove this useless assignment to local variable 'testQuestionObj'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1854)

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 9]
```diff
@@ -101,5 +111,78 @@ public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util)
...
+                await _dbContext.SaveChangesAsync();
+                return _stringConstants.SuccessfullySaved;
+            }
+        }
+
+        public async Task<TestAC> GetTestDetailsByIdAsync(int testId)
+        {
+            TestAC testAcObject = new TestAC();
+            var test = await _dbContext.Test.FindAsync(testId);
+            testAcObject = Mapper.Map<Test, TestAC>(test);
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: MAJOR Remove this useless assignment to local variable 'testAcObject'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1854)

**Comment by chintans** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs [hunk 10]
```diff
@@ -26,13 +28,6 @@ public interface ITestsRepository
...
-        /// Gets the settings saved of a particular Test
-        /// </summary>
-        /// <param name="id">The parameter "id" is used to get the Settings of a Test by its Id</param>
-        /// <returns>Settings Saved for the selected Test</returns>
-        Task<Test> GetTestByIdAsync(int id);
-
-        /// <summary>
         /// Updates the changes made to the settings of a Test
         /// </summary>
         /// <param name="testObject">The parameter "testObject" is an object of Test</param>
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: I dont find any testObject as method parameter.

File: Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs [hunk 11]
```diff
@@ -67,5 +62,28 @@ public interface ITestsRepository
...
         /// <returns>Boolean:true if an attendee exist or else false</returns>
         Task<bool> IsTestAttendeeExistAsync(int id);
+
+        /// <summary>
+        /// Gets all the questions contained by a particular category in a test
+        /// </summary>
+        /// <param name="testId">Parameter "testId" takes value of a Test's Id from route </param>
+        /// <param name="CategoryId">Parameter "CategoryId" takes value of a Category's Id from route</param>
+        /// <returns>List of QuestionAC objects of particular category</returns>      
+        Task<List<QuestionAC>> GetAllTestCategoryQuestionsByIdAsync(int testId, int CategoryId);
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: improper naming

File: Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs [hunk 12]
```diff
@@ -67,5 +62,28 @@ public interface ITestsRepository
...
+
+        /// <summary>
+        /// Adds selected question(s) to the particular "Test" 
+        /// </summary>
+        /// <param name="QuestionsToAddTest">Parameter "QuestionsToAddTest" is a list of questions to be added to Test which takes value from body</param>
+        /// <param name="testId">Parameter "testId" takes value of a Test's Id from route</param>
+        /// <returns>A string message of Success</returns>
+        Task<string> AddTestQuestionsAsync(List<QuestionAC> QuestionsToAddTest, int testId);
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: improper naming

File: Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs (hunk 10 above)
Comment (asif-khan17) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: This is madhurima's method and she  has used this object

Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: if it is not needed delete it else keep it

Comment (Madhurima94) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: Mam, I need this object to update the fields of Test Settings Page in database. So, it is to be kept.

**Pull request closed without merging**